---
title: "Proteomics final assignment"
author: "Margarita Onvumere"
date: "2024-06-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

```{r, include=FALSE}
# if (!require("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# 
# BiocManager::install("limma")
# BiocManager::install("NMF")
# BiocManager::install("impute")

library(readxl)
library(limma)
library(ape)
library(dendextend)
library(RColorBrewer)
library(pvclust)
library(gplots)
library(NMF)
library(vegan)
library(Biobase)
library(DT)
library(ggplot2)
library(impute)
library(ggrepel)
```

# Uploading dataset and EDA

```{r}
data <- read.csv('peaks_data.csv')
nrow(data)

# rownames(data) <- data$Gene_id
# Error in `.rowNamesDF<-`(x, value = value) : 
#   duplicate 'row.names' are not allowed
# In addition: Warning message:
#   non-unique values when setting 'row.names': ‘’, ‘GNAS’, ‘NACA’, ‘nan’, ‘PALM2AKAP2’ 

data_rows_check <- data[data$Gene_id %in% c("NACA", 'PALM2AKAP2', "GNAS"), ]
# For GNAS - deleting a duplicate
# PALM2AKAP2 and NACA - selecting a row with less NAs

data <- data[-c(849, 1322, 3431),]
data <- data[!(data$Gene_id %in% c("", "nan")), ]
rownames(data) <- data$Gene_id
data <- data[,-c(1,2,3)]

# Remove genes with half and more missing values
genes_with_NA <- names(which(rowSums(is.na(data)) > ncol(data)/2))
data_filt <- data[!rownames(data) %in% genes_with_NA,]

## Filling NAs
data_filt <- as.matrix(data_filt)
data_filt_trans <- t(data_filt)
knn_data_filt <- impute.knn(data_filt_trans, k = 5)
knn_data_filt <- knn_data_filt$data
knn_data_filt <- as.data.frame(knn_data_filt)
knn_data_filt <- t(as.matrix(knn_data_filt))
any(is.na(knn_data_filt))

data_experiment <- as.factor(c(rep("1", 14), rep("2", 19)))
```

# Data normalisation

```{r}
pal <- brewer.pal(n = 9, name = "Set1")
cols <- pal[data_experiment]
boxplot(knn_data_filt, outline = FALSE, main = "Raw data", col = cols)
legend("topright", levels(data_experiment), fill = pal, bty = "n", xpd = T)

data_filt_knn_log <- log1p(knn_data_filt)
data_norm <- normalizeQuantiles(as.matrix(data_filt_knn_log))
boxplot(data_norm, outline = FALSE, main = "Normalised data (log(1+x))", col = cols)
legend("topright", levels(data_experiment), fill = pal, bty = "n", xpd = T)
any(is.na(data_norm))
```

# PCA

```{r}
data_pca <- t(data_norm)
terapod_pca <- rda(data_pca, scale = TRUE)
data_scores <- data.frame(data_pca,
                        scores(terapod_pca, display = "sites", choices = c(1, 2, 3), scaling = "sites"))
p_scores <- ggplot(data_scores, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = rownames(data_pca)), alpha = 0.5) +
  coord_equal(xlim = c(-3, 3), ylim = c(-3, 3)) + ggtitle(label = "Ordination in the axes of principal components") + theme_bw()
p_scores
```

# MA-plot

```{r}
maplot <- function(X1, X2, pch = 21, main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", lpars = list(col = "blue", lwd = 2), ...){
  X <- (rowMeans(X2) + rowMeans(X1)) / 2
  Y <- rowMeans(X2) - rowMeans(X1)
  scatter.smooth(x = X, y = Y, main = main, pch = pch, xlab = xlab, ylab = ylab, lpars = lpars, ...)
  abline(h = c(-1, 0, 1), lty = c(2, 1, 2))
}

maplot(data_norm[,1:14], data_norm[,15:33])

```

# Differential expression, heatmap

```{r}
expr_data <- as.matrix(data_norm)
pheno_data <- data.frame(data_experiment)
rownames(pheno_data) <- colnames(data_norm)
pheno_metadata <- data.frame(
  labelDescription = c("Experimental condition"),
  row.names=c("Condition"))
pheno_data <- new("AnnotatedDataFrame",
                  data = pheno_data,
                  varMetadata = pheno_metadata)
feature_data <- data.frame(Prot = rownames(expr_data))
rownames(feature_data) <- rownames(expr_data)
feature_metadata <- data.frame(
  labelDescription = c("Protein name"),
  row.names = c("Protein"))
f_data <- new("AnnotatedDataFrame",
              data = feature_data,
              varMetadata = feature_metadata)
exp_set <-
  ExpressionSet(assayData = expr_data,
                phenoData = pheno_data,
                featureData = f_data)

X <- model.matrix(~ data_experiment, pData(exp_set))
fit <- lmFit(exp_set, design = X, method = "robust", maxit = 1000)
efit <- eBayes(fit)

MA_limma <- function(efit, coef, n = 10, signif = TRUE, fdr = 0.05, lfc = 0, text = TRUE, cex.text = 0.8, col.text = "grey20", main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", pch = 19, pch.signif = 21, col = "darkgreen", alpha = 0.3, cex = 0.3, ...){
  R <- efit$coefficients[, coef]
  I <- efit$Amean
  col_btransp <- adjustcolor(col, alpha.f = alpha)
  plot(I, R, cex = cex, main = main, pch = pch, xlab = xlab, ylab = ylab, col = col_btransp, ...)
  abline(h = 0)
  if(signif){
    sign <- p.adjust(efit$p.value[, coef], method = "BH") <= fdr
    large <- abs(efit$coefficients[, coef]) >= lfc
    points(I[sign & large], R[sign & large], cex = cex*2, col = "orange2", pch = pch.signif)
  }
  if(text){
    ord <- order(efit$lods[, coef], decreasing = TRUE)
    top_n <- ord[1:n]
    text(I[top_n], R[top_n], labels = efit$genes[top_n, ], pos = 4, cex = cex.text, col = col.text)
  }
}
MA_limma(efit, coef = 2, n = 30)


my_list <- topTable(efit, coef = 2, n = 20)
dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
dat <- as.matrix(exprs(dif_exp_set))

topTable(efit, coef = 2)
numGenes <- nrow(exprs(exp_set))
full_list <- topTable(efit, number = numGenes)
full_list <- full_list[full_list$adj.P.Val <= 0.05,]
write.csv(full_list, 'full_list.csv')

dat <- as.matrix(exprs(dif_exp_set))
pal_blue_red <- colorpanel(75, low = "steelblue", mid = "black", high = "red")

heatmap.2(dat, col = pal_blue_red, scale = "row", key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.9, cexCol = 1, margins = c(4, 3), keysize = 0.8, key.par = list(mar = c(2,1,2,1)))


MA_limma(efit, coef = 2, n = 4)
MA_limma(efit, coef = 2, n = 20, text = F, lfc = 1)

```

# Volcano plot

```{r}
volcano_list <- full_list
volcano1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()
volcano2 <- volcano1 + geom_vline(xintercept=c(-0.6, 0.6), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")
volcano_list$diffexpressed <- "NO"       

volcano_list$diffexpressed[volcano_list$logFC > 1 & volcano_list$adj.P.Val < 0.05] <- "UP"
volcano_list$diffexpressed[volcano_list$logFC < -1 & volcano_list$adj.P.Val< 0.05] <- "DOWN"
volcanodif1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed)) + geom_point() + theme_minimal()
volcanodif2 <- volcanodif1 + geom_vline(xintercept=c(-1, 1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")
volcano_list$Prot <- as.character(volcano_list$Prot)
volcano_list$delabel <- NA
volcano_list$delabel[volcano_list$diffexpressed != "NO"] <- volcano_list$Prot[volcano_list$diffexpressed != "NO"]

plot_proteins <- ggplot(data=volcano_list, aes(x = logFC, y=-log10(adj.P.Val), col=diffexpressed, label=delabel)) + 
  geom_point() + 
  theme_minimal() +
  geom_text_repel(size=3, colour = 'black', max.overlaps = 30)

plot_final <- plot_proteins + geom_vline(xintercept=c(-1, 1), col="red") + geom_hline(yintercept=-log10(0.05), col="red")
plot_final

```

# Conclusions

Genes that are upregulated in people with cardiac valve calcification are associated with several  biological pathways, primarily related to inflammation and cellular stress responses, which are closely linked to the calcification process.

Based on the PCA plot, we don't see a batch effect for our data.
